flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mailbox.java

```diff
@@ -20,17 +20,85 @@
 
 import javax.annotation.Nonnull;
 
+import java.util.List;
+
 /**
  * A mailbox is basically a blocking queue for inter-thread message exchange in form of {@link Runnable} objects between
- * multiple producer threads and a single consumer.
+ * multiple producer threads and a single consumer. This has a lifecycle of closed -> open -> (quiesced) -> closed.
  */
 public interface Mailbox extends MailboxReceiver, MailboxSender {
 
 	/**
-	 * The effect of this is that all pending letters are dropped and the given priorityAction
-	 * is enqueued to the head of the mailbox.
+	 * This enum represents the states of the mailbox lifecycle.
+	 */
+	enum State {
+		OPEN, QUIESCED, CLOSED
+	}
+
+	/**
+	 * Open the mailbox. In this state, the mailbox supports put and take operations.
+	 */
+	void open();
+
+	/**
+	 * Quiesce the mailbox. In this state, the mailbox supports only take operations and all pending and future put
+	 * operations will throw {@link MailboxStateException}.
+	 */
+	void quiesce();
+
+	/**
+	 * Close the mailbox. In this state, all pending and future put operations and all pending and future take
+	 * operations will throw {@link MailboxStateException}. Returns all letters that were still enqueued.
+	 *
+	 * @return list with all letters that where enqueued in the mailbox at the time of closing.
+	 */
+	@Nonnull
+	List<Runnable> close();
+
+	/**
+	 * The effect of this is that all pending letters in the mailbox are dropped and the given priorityLetter
+	 * is enqueued to the head of the mailbox. Dropped letters are returned. This method should only be invoked
+	 * by code that has ownership of the mailbox object and only rarely used, e.g. to submit special events like
+	 * shutting down the mailbox loop.
+	 *
+	 * @param priorityLetter action to enqueue atomically after the mailbox was cleared.
+	 * @throws MailboxStateException if the mailbox is quiesced or closed.
+	 */
+	@Nonnull
+	List<Runnable> clearAndPut(@Nonnull Runnable priorityLetter) throws MailboxStateException;
+
+	/**
+	 * Adds the given action to the head of the mailbox. This method will block if the mailbox is full and
+	 * should therefore only be called from outside the mailbox main-thread to avoid deadlocks.
+	 *
+	 * @param priorityLetter action to enqueue to the head of the mailbox.
+	 * @throws InterruptedException on interruption.
+	 * @throws MailboxStateException if the mailbox is quiesced or closed.
+	 */
+	void putFirst(@Nonnull Runnable priorityLetter) throws InterruptedException, MailboxStateException;
```

Comment:
If every method enqueuing the `Runnable` returned a `CompletableFuture<?>`, we would have a better mechanism of canceling them, instead of instance of checking the `Runnable`, right? Plus it would give us a generic mechanism for checking whether an action has completed or not (especially useful in tests). 

Are there some drawbacks of using `CompletableFuture`?

Reply:
I was thinking about that and decided against it for a couple of reasons. First, overhead is introduced even for caller that do not care about completion, cancelation etc., which seems to be most cases. Second, you can always build this on top and also the user-facing fassade of the executor service has methods that return `Future`. If you wonder why not `CompletableFuture` instead, it is because the user can always create such futures themselves via `CompletableFuture.runAsync(Runnable, Executor)`. However, there is a downside to this.
Such calls enqueue a special instance of `RunnableFuture` under the hood, but calling cancel on them is never propagated back to the `CompletableFuture`. That is just a general shortcoming of this implementation in the JDK in my opinion. So, if you want to also propagate cancellation to completable futures, that would mean you have to remember all of them, in parallel to the `Runnable`s, because they are not `Runnable`. So there are many options already possible with this interface that does not impose return values on all clients:
- no feedback
- Future
- compleable future via `runAsync`
- your own feedback mechanism if you submit instances of `Runnable(Future)`